README.md
<a href="https://scan.coverity.com/projects/steppenwolfe65-cex">
  <img alt="Coverity Scan Build Status" src="https://scan.coverity.com/projects/14233/badge.svg"/>
</a>

# CEX

The CEX Cryptographic library in C++

## Intro
### Welcome
CEX is being written from the ground-up as a powerful and accessible post-quantum secure cryptographic library; a flexible security model, performance-oriented, designed for ease of use, and the automation of complex tasks.

There are some new ideas, and new technologies, as I intend to push the envelope, and so am authoring this with a determination to make the fastest, most intuitive, most secure implementations possible.

This library is being built in two stages; the symmetric cryptography, which consists of ciphers, hash functions, MACs, RNGs, TRNGs etc, preliminary work has been completed as of version v1.0. 
That work is still evolving however, as improvements and additions to the symmetric cryptography will continue throughout the libraries lifetime.
The second half is the the addition of asymmetric cryptography, with a strong focus on post-quantum security. This work is well under way, and this release contains the ModuleLWE (Kyber) and McEliece (Niederreiter) asymmetric ciphers, as well as the Dilithium, XMSS, and SPHINCS+ signature schemes.

Though efforts have been made towards future compatibility with other platforms, this is currently a Windows-only library, but is currently scheduled for multi-platform compatibility (Android, Linux, and possibly iOS) by the spring of 2023.
This has been tested on MSVC-2017, MSVC-2019, and MSVC-2022 in ARM/x86/x64 debug and release modes, using the MSVC and Intel compiler tool-chains, and the MacOS and Linux operating systems. This project optionally uses OpenMP support.

Works with or without intrinsics, set the test project and the library to a supported instruction set to test the intrinsics enhanced implementations.
The SIMD support is set to AVX2 by default, (AVX implementations are also in place on some ciphers and protocols, set the enhanced instruction flag to your system supported set: arch:AVX2, or the minimum supported instructions arch:AVX, or None, arch:IA32). The library also has AVX512 support, that can be enabled through the CexConfig.h file.

The Win folder contains a visual studio test project, which tests each cipher and protocol with it's official test vectors, and has a set of digest and symmetric cipher speed tests. Make sure the project properties SIMD and OpenMP support are enabled before running the project, and for speed tests, compile in release mode.
If running the executable, the Win\Test\Vectors folder containing the cipher KAT files must be in the executables path.

For more information on the current capabilities of the library, refer to the libraries help pages.

### New in Current Release 1.1.0.0 (version A10):
* Rainbow, NTRU Prime, and New Hope have been removed.
* All remaining implementations have been updated to the NIST PQ Round 3 final versions
* Asymmetric primitives that use AVX instructions have been added

### New in Current Release 1.0.0.8 (version A8)
* The Elliptic Curve Diffie Hellman Key exchange (EC25519)
* The Elliptic Curve Digital Signature Algorithm (ED25519)
* Integration of AES-NI 256 and 512-bit instructions
* The Rainbow signature scheme
* The 512-bit block Rijndael authenticated stream cipher RWS (end of the world cipher)
* The ChaCha derivitive CSX-512 authenticated stream cipher using 64-bit integers and a 1024-bit block, and a 512-bit key
* Changes to extended 1024-bit versions of KMAC, SHA3, and SHAKE
* The integration of SecureVector (memory locked arrays) throughout
* The addition of an IPv4/IPv6  networking stack
* The addition of a Keccak-based passphrase based KDF: SCBKDF
* The addition of a new hash-based AEAD mode: HBA
* Asymmetric ciphers and signature schemes updated to NIST-PQ Round 2 versions (updated again in November after the Round-3 versions upload)

## Contents
### Asymmetric Ciphers
* The NTRU asymmetric cipher (S-Prime and L-Prime)
* The RingLWE asymmetric cipher (New Hope -N1024/N2048)
* The Niederreiter dual form of the McEliece cipher
* The ModuleLWE asymmetric cipher (Kyber)

### Asymmetric Signature Schemes
* The Rainbow signature scheme
* The Dilithium asymmetric signature schemes
* The SPHINCS+ 256F-SHAKE128/256 asymmetric signature schemes
* The XMSS/XMSS-MT asymmetric signature schemes

### Block Ciphers
Note: Each cipher can be deployed as either the standard form (AES, Serpent), or the extended hybrid using cSHAKE or HKDF(SHA2) key expansion
* The AES-NI Hash eXtended cipher (AHX)
* The fallback Rijndael Hash eXtended cipher (RHX)
* The Serpent Hash eXtended cipher (SHX)

### Block Cipher Modes
* The Hash Based Authentication mode (HBA)
* Galois Counter authenticated block cipher Mode (GCM)
* Cipher Block Chaining mode (CBC)
* Cipher FeedBack mode (CFB)
* Big-Endian integer Counter mode (CTR)
* Electronic CodeBook mode (ECB)
* Little-Endian Integer Counter Mode (ICM)
* Output FeedBack Mode (OFB)

### Block Cipher Padding
* The ISO7816 Padding Scheme
* The PKCS7 Padding Scheme
* The Trailing Bit Compliment Padding Scheme (TBC)
* The X.923 Padding Scheme

### Stream Ciphers 
Note: Integrated an optional built-in authentication generator (KMAC) to each stream cipher
* The Authenticate and Encrypt and AEAD wide-block Rijndael-256 stream cipher implementation (ACS/RCS).
* The RWS authenticated stream cipher: Rijndael with a 512-bit block, running 40/80 rounds in an KMAC authenticated counter-mode stream cipher
* Threefish 256/512/1024 authenticated stream ciphers
* ChaCha256-P20 and the [experimental] CSX512 authenticated stream ciphers

### Message Digests
Note: Every message digest implementation has both parallel and sequential modes of operation
* The Blake2 256 and 512 bit variants (Blake256/Blake512)
* The SHA-3 256 and 512 bit variants (Keccak256/Keccak512)
* The SHA-2 256 and 512 bit variants (SHA256/SHA512)
* The Skein 256, 512,bit variants(Skein256/Skein512/Skein1024)

### DRBGs
* The Block cipher Counter mode Generator using the wide-block Rijndael-256 (BCG)
* The custom cSHAKE Generator (CSG)
* The HMAC Counter Generator (HCG)

### KDFs
* Hash based Key Derivation Function (HKDF)
* Key Derivation Function Version 2 (KDF2)
* Passphrase Based Key Derivation Version 2 (PBKDF2)
* The SHAKE cost based passphrase generator (SCBKDF)
* The 128/256/512 SHAKE XOF function

### MACs
* Cipher based Message Authentication Code generator (CMAC)
* Galois/Counter Message Authentication Code generator (GMAC)
* Hash based Message Authentication Code generator (HMAC)
* The Poly1305 Message Authentication Code generator (Poly1305)
* Keccak based Message Authentication Code generator (KMAC)

### PRNGs
* The auto-seeded Block cipher Counter mode Rng (BCR)
* The auto-seeded message Digest Counter Generator (DCR)
* The auto-seeded HMAC Counter Generator (HCR)
* An implementation of a Passphrase Based PRNG (PBR)
* The prng extension wrapper class (SecureRandom)

### Entropy Providers
* Auto seed Collection Provider (ACP)
* CPU Jitter entropy Provider (CJP)
* Local Crypto Service Provider (CSP)
* System Entropy Collection Provider (ECP)
* Intel RdRand/RdSeed Provider (RDP)

## License
This project is licensed under the AGPL version 3 (AGPLv3):
https://www.gnu.org/licenses/gpl-3.0.en.html

## Disclaimer
This project contains strong cryptography, before downloading the source files, 
it is your responsibility to check if the extended symmetric cipher key lengths (512 bit and higher), and other cryptographic algorithms contained in this project are legal in your country. 
If you use this code, please do so responsibly and in accordance to law in your region.


History.txt

## Trajectory
The current version is <B>1.1.0.0</B> (A10 version), which are the major, minor, patch, and release codes.
All asymmetric primitives have been updated to NIST PQ Round 3 final versions

### New in Current Release 1.1.0.0 (version A10):
* Rainbow, NTRU Prime, and New Hope have been removed.
* All remaining implementations have been updated to the NIST PQ Round 3 final versions
* Asymmetric primitives that use AVX instructions have been added

### New in Current Release 1.0.0.8 (version A8):
* The Rainbow signature scheme
* The 512-bit block Rijndael authenticated stream cipher RWS
* The ChaCha derivitive CSX-512 authenticated styream cipher using 64-bit integers and a 1024-bit block
* Changes to extended 1024-bit versions of KMAC, SHA3, and SHAKE
* The integration of SecureVector (memory locked arrays) throughout
* The addition of an IPv4/IPv6  networking stack
* The addition of a Keccak-based passphrase based KDF: SCBKDF
* The addition of a new hash-based AEAD mode: HBA
* Asymmetric ciphers and signature schemes updated to NIST-PQ Round 2 versions (updated again in November after the Round-3 versions upload)

### New in Current Release 1.0.0.7 (version A7):
* The Dilithium asymmetric signature schemes
* The SPHINCS+ 256F-SHAKE128/256 asymmetric signature schemes
* The XMSS/XMSS-MT asymmetric signature schemes
* The NTRU Prime asymmetric cipher
* The cSHAKE XOF function added as an HX symmetric cipher key expansion option
* Update of all base asymmetric ciphers and signature schemes to match NIST PQ entries
* Message digests updated for performance and security
* Integrated an optional built-in authentication generator (HMAC/KMAC) to each stream cipher
* Added the Authenticate and Encrypt and AEAD stream cipher implementation (Authenticated Cipher Stream: ACS).
* Threefish 256/512/1024 authenticated stream ciphers
* ChaCha256-P20 and ChaCha512-P80 authenticated stream ciphers

### Planned Release 1.0.0.8 (version A8):
## ETA is August 1st, 2019
* Authenticated stream ciphers ACS,ThreeFish-256/512/1024, and ChaCha256/512
* Update secure key mechanisms and integrate internally
* Full security, documentation, and code review cycles
* Begin optimization and review of asymmetric primitives
* Create accompanying website and publish documentation

### Planned Release 1.2.0.1
* DLL API (.NET/Java)
* Direct Trust Model DTM-KEX (external?)
* Post-Quantum Secure DNS, PQ-SECDNS (external)
* Shared Trust Model STM-KEX (external?)

### Planned Release 1.1.0.1
* RingLWE
* McEliece
* ModuleLWE
* NTRU Prime
* Picnic
* SPHINCS+
* Dilithium
* XMSS/XMSS-MT

### Planned Release 1.2.0.1
* PQ-TLS
* STM-KEX
* Android/iOS/Linux compatibility
* DLL API

## History
  
### Version 1.0.0.7q, April 07, 2019
* The Dilithium asymmetric signature scheme
* The SPHINCS+ asymmetric signature scheme
* The NTRU Prime asymmetric cipher
* The cSHAKE XOF function added as an HX symmetric cipher key expansion option
* Update of all base asymmetric ciphers to match NIST PQ entries
* Message digests updated for performance and security
* Integrated an optional built-in authentication generator (HMAC/KMAC) to each stream cipher
* Added the Authenticate and Encrypt and AEAD stream cipher implementation (Authenticated Cipher Stream: ACS).
* Threefish 256/512/1024 authenticated stream ciphers
* ChaCha256-P20 and ChaCha512-P80 authenticated stream ciphers

### Version 1.0.0.5, March 16, 2018
* The library is now Misra C++ 2014 compliant
* Added the ModuleLWE asymmetric cipher
* Added the SHAKE Key Derivation Function
* Added the cSHAKE generator (CSG)
* Added the KMAC message authentication code generator
* Addition of asymmetric cipher Encapsulate/Decapsulate api

### Version 1.0.0.4, November 11, 2017
* Added the McEliece asymmetric cipher
* Added the 1024bit Keccak digest
* Added the Poly1305 Message Authentication Code generator
* Many changes made towards future SEI-CERT and Misra compliance (Full Misra compliance by 1.0.0.5)
* Full coding standards sweep of the library
* Integration of std::array and std::unique_ptr throughout
* A full optimization cycle completed

### Version 1.0.0.3, June 30, 2017
* Added asymmetric cipher interfaces and framework
* Added RingLWE asymmetric cipher
* Added the Auto Collection seed Provider (ACP)
* Addition of the HCR prng
* Renaming of the drbgs to xCG format: BCG, DCG, and HCG; Block cipher Counter Generator, Digest and HMAC Counter Generators
* Overhaul of SecureRandom and prng classes

### Version 1.0.0.2: April 23, 2017
Last of 1.0 sweep of the symmetric library before the second half of the project engages, with thousands of changes made throughout, and the addition of (!experimental) AVX512 support.
* Added a vectorized MemUtils class, with SIMD 128/256/512 copy, clear, set-value, and xor functions.
* Integrated vectorized replacements for memcpy, xor, and memset throughout, including cipher mode support for AVX512, (I don't have a xeon to test this, maybe you can help?).
* Reformatting of headers (inline accessors removed and the override hint added).
* Many small TODOs finished, api synchronized, and formatting and documentation changes throughout.

### Version 1.0: March 28, 2017

CEX/ACS.h
// The GPL version 3 License (GPLv3)
// 
// Copyright (c) 2020 vtdev.com
// This file is part of the CEX Cryptographic library.
// 
// This program is free software : you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program. If not, see <http://www.gnu.org/licenses/>.
//
//
// Implementation Details:
// An implementation of an AES-NI Rijndael-256 authenticated Counter Mode (ACS).
// Version 1.0d
// Written by John G. Underhill, March 8, 2019
// Updated January 09, 2020
// Updated July 18, 2020
// Contact: develop@vtdev.com

#ifndef CEX_ACS_H
#define CEX_ACS_H

#include "IStreamCipher.h"
#include "IMac.h"
#include "Intrinsics.h"
#include "KmacModes.h"

NAMESPACE_STREAM

using Mac::IMac;
using Enumeration::KmacModes;

/// <summary>
/// An AES-NI implementation of the Rijndael symmetric 256-bit block-cipher, operating in a Little-Endian counter-mode, as an Authenticate, Encrypt, and Additional Data (AEAD) stream cipher implementation  (AESNI-256 Cipher Stream).
/// <para>This cipher uses an optional authentication mode; Poly1305, HMAC(SHA2), or KMAC set through the constructor to authenticate the stream.</para>
/// </summary> 
/// 
/// <example>
/// <description>Encrypting an array of bytes:</description>
/// <code>
/// SymmetricKey kp(Key, Nonce);
/// // initialize the Rijndael cipher with the SHAKE-256 key-schedule extension
/// ACS cipher(StreamAuthenticators::KMAC256);
/// // mac code is appended to the cipher-text stream in authentication mode
/// cipher.Initialize(true, kp);
/// cipher.Transform(Input, InOffset, Output, OutOffset, Length);
/// </code>
///
/// <description>Decrypt and verify an array:</description>
/// <code>
/// SymmetricKey kp(Key, Nonce);
/// ACS cipher(StreamAuthenticators::KMAC256);
/// // initialize for decryption
/// cipher.Initialize(false, kp);
///
/// // decrypt the ciphertext and catch authentication failures
/// try
/// {
///		cipher.Transform(Input, InOffset, Output, OutOffset, Length);
/// }
/// catch (CryptoAuthenticationFailure)
/// {
///		// authentication has failed, do something...
/// }
/// </code>
/// </example>
/// 
/// <remarks>
/// <description><B>Overview:</B></description>
/// <para>
/// The ACS symmetric cipher is a 256-bit wide-block Rijndael based, Authenticate, Encrypt, and Additional Data (AEAD) authenticated stream-cipher. \n
/// ACS has provable security, based on the Rijndael-256 permutation function used by this cipher. \n
/// ACS first encrypts the plain-text using a Little-Endian counter mode, then processes that cipher-text using a MAC function used for data authentication. \n\n
/// When each transform encryption call is completed, the MAC code is generated and appended to the output vector automatically. \n
/// Decryption performs these steps in reverse, processing the cipher-text bytes through the MAC function, and if authentication succeeds, then decrypting the data to plain-text. \n
/// During decryption, if the MAC codes do not match, a CryptoAuthenticationFailure exception error is thrown.</para>
///
/// <description><B>Multi-Threading:</B></description>
/// <para>The encryption and decryption functions of the ACS mode can be multi-threaded. This is achieved by processing multiple blocks of message input independently across threads. \n
/// The ACS stream cipher also leverages SIMD instructions to 'double parallelize' those segments. An input block assigned to a thread
/// uses SIMD instructions to decrypt/encrypt blocks in parallel, depending on which framework is runtime available, AVX, AVX2, or AVX512 SIMD instructions. \n
/// Input blocks equal to, or divisble by the ParallelBlockSize() are processed in parallel on supported systems.
/// The cipher transform is parallelizable, however the authentication pass, is processed sequentially.</para>
///
/// <description>Implementation Notes:</description>
/// <list type="bullet">
/// <item><description>Supported key sizes are 32, 64, and 128 bytes (256, 512, and 1024 bits).</description></item>
/// <item><description>The required Nonce size is 32 bytes (256 bits).</description></item>
/// <item><description>The ISymmetricKey Info value can be used as a cipher tweak to create a unique ciphertext and MAC output.</description></item>
/// <item><description>The ciphers Initialize function can use either a SymmetricKey, or an encrypted SymmetricSecureKey key container.</description></item>
/// <item><description>The internal block input-size is fixed at 32 bytes wide (256 bits).<</description></item>
/// <item><description>This cipher is capable of authentication by setting the constructors StreamAuthenticators enumeration to Poly1305, or one of the HMAC or KMAC options.</description></item>
/// <item><description>In authentication mode, during encryption the MAC code is automatically appended to the output cipher-text, during decryption, this MAC code is checked and authentication failure will generate a CryptoAuthenticationFailure exception.</description></item>
/// <item><description>If authentication is enabled, the cipher and MAC keys are generated by passing the input cipher-key through an instance of cSHAKE, this will yield a different cipher-text output from non-authenticated modes.</description></item>
/// <item><description>In authenticated mode, the internal keys generated by cSHAKE will be unique with each MAC generator, each authentication option should be considered a seperate and distinct variant of the cipher, for example, RCS256-SHAKE512 or RCS256-HMAC512</description></item>
/// <item><description>The Info string is optional, but can be used to create a tweakable cipher, this can be used for adding additional key material, or using a second key to restrict decryption to a domain based system.</description></item>
/// <item><description>Permutation rounds are fixed 22, 30, and 38, for 256, 512, and 1024-bit keys.</description></item>
/// <item><description>Authentication using Poly1305, HMAC, or KMAC, can be invoked by setting the StreamAuthenticators parameter in the constructor, when set to None, authentication is disabled.</description></item>
/// <item><description>The class functions are virtual, and can be accessed from an IStreamCipher instance.</description></item>
/// <item><description>The transformation methods can not be called until the Initialize(ISymmetricKey) function has been called.</description></item>
/// <item><description>Encryption can both be pipelined (AVX, AVX2, or AVX512), and multi-threaded with any even number of threads, the configuration can be modified using the ParallelProfile() accessor function.</description></item>
/// <item><description>If the system supports Parallel processing, and ParallelProfile().IsParallel() is set to true; passing an input block of ParallelProfile().ParallelBlockSize() to the transform will be auto-parallelized.</description></item>
/// <item><description>The ParallelProfile().ParallelThreadsMax() property is used as the thread count in the parallel loop; it defaults to the maximum number of available virtual cores, but is user-assignable, and must be an even number no greater than the number of processer cores on the system.</description></item>
/// <item><description>ParallelProfile().ParallelBlockSize() is calculated automatically based on processor(s) cache size but can be user defined, but must be evenly divisible by ParallelProfile().ParallelMinimumSize().</description></item>
/// <item><description>The ParallelBlockSize(), IsParallel(), and ParallelThreadsMax() accessors, can be changed through the ParallelProfile() property, but this has been auto-calculated based on the systems hardware, modifications are not recommended</description></item>
/// </list>
///
/// <description>Guiding Publications:</description>
/// <list type="number">
/// <item><description>NIST <a href="http://csrc.nist.gov/publications/fips/fips197/fips-197.pdf">AES Fips 197</a>.</description></item>
/// <item><description>NIST <a href="http://csrc.nist.gov/archive/aes/rijndael/Rijndael-ammended.pdf">Rijndael ammended</a>.</description></item>
/// <item><description>HMAC <a href="http://tools.ietf.org/html/rfc2104">RFC 2104</a>.</description></item>
/// <item><description>Fips <a href="http://csrc.nist.gov/publications/fips/fips198-1/FIPS-198-1_final.pdf">198.1</a>.</description></item>
/// <item><description>HKDF <a href="http://tools.ietf.org/html/rfc5869">RFC 5869</a>.</description></item>
/// <item><description>NIST <a href="http://csrc.nist.gov/publications/drafts/800-90/draft-sp800-90b.pdf">SP800-90B</a>.</description></item>
/// <item><description>SHA3 <a href="http://keccak.noekeon.org/Keccak-submission-3.pdf">The Keccak digest</a>.</description></item>
/// <item><description>FIPS 202: <a href="http://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.202.pdf">Permutation Based Hash</a> and Extendable Output Functions</description></item>
/// <item><description>NIST <a href="http://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-185.pdf">SP800-185</a> SHA-3 Derived Functions.</description></item>
/// <item><description>Team Keccak <a href="https://keccak.team/index.html">Homepage</a>.</description></item>
/// </list>
/// </remarks>
class ACS final : public IStreamCipher
{
private:

	static const size_t BLOCK_SIZE = 32;
	static const size_t IK256_SIZE = 32;
	static const size_t IK512_SIZE = 64;
	static const size_t IK1024_SIZE = 128;
	static const size_t INFO_SIZE = 16;
	static const size_t MAX_PRLALLOC = 100000000;
	// Transformation round counts per input key size:
	// modifying these values will increase the rounds processed by the cipher.
	// These are the minimum sizes, changes will cause test failures,
	// and incompatibility with standard the version.
	static const size_t RK256_COUNT = 22;
	static const size_t RK512_COUNT = 30;
	static const size_t RK1024_COUNT = 38;
	static const size_t STATE_PRECACHED = 2048;
	static const size_t STATE_THRESHOLD = 838;
	static const byte UPDATE_PREFIX = 0x80;


#if defined(CEX_HAS_AVX512)
	static const __m512i NI512K0;
	static const __m512i NI512K1;
#endif
#if defined(CEX_EXTENDED_AESNI)
	static const __m256i NI256K0;
	static const __m256i NI256K1;
#else
	static const __m128i NIBMASK;
	static const __m128i NISMASK;
#endif

	class AcsState;
	std::unique_ptr<AcsState> m_acsState;
	std::unique_ptr<IMac> m_macAuthenticator;
	ParallelOptions m_parallelProfile;

public:

	//~~~Constructor~~~//

	/// <summary>
	/// Copy constructor: copy is restricted, this function has been deleted
	/// </summary>
	ACS(const ACS&) = delete;

	/// <summary>
	/// Copy operator: copy is restricted, this function has been deleted
	/// </summary>
	ACS& operator=(const ACS&) = delete;

	/// <summary>
	/// Initialize the Cipher Mode using a block cipher type name.
	/// <para>The cipher instance is created and destroyed automatically.</para>
	/// </summary>
	///
	/// <param name="Authenticate">Activate the authentication option</param>
	///
	/// <exception cref="CryptoSymmetricException">Thrown if an invalid authentication type is chosen</exception>
	explicit ACS(bool Authenticate);

	/// <summary>
	/// Initialize the stream cipher using a secure-vector serialized state.
	/// <para>The Serialize function stores the internal state of the cipher, so that it can be reinitialized,
	/// without the need to call the Initialize function and key-schedule. 
	/// If this constructor is used, the cipher is fully initialized to the values it had when the Serialize function was called.</para>
	/// </summary>
	///
	/// <param name="State">The serialized state, created by the Serialize() function</param>
	///
	/// <exception cref="CryptoSymmetricException">Thrown if an invalid state array is used</exception>
	explicit ACS(SecureVector<byte> &State);

	/// <summary>
	/// Destructor: finalize this class
	/// </summary>
	~ACS() override;

	//~~~Accessors~~~//

	/// <summary>
	/// Read Only: The stream ciphers type name
	/// </summary>
	const StreamCiphers Enumeral() override;

	/// <summary>
	/// Read Only: Cipher has authentication enabled
	/// </summary>
	const bool IsAuthenticator() override;

	/// <summary>
	/// Read Only: The cipher has been initialized for encryption
	/// </summary>
	const bool IsEncryption() override;

	/// <summary>
	/// Read Only: Cipher is ready to transform data
	/// </summary>
	const bool IsInitialized() override;

	/// <summary>
	/// Read Only: Processor parallelization availability.
	/// <para>Indicates whether parallel processing is available with this cipher.
	/// If parallel capable, input/output data arrays passed to the transform must be ParallelBlockSize in bytes to trigger parallelization.</para>
	/// </summary>
	const bool IsParallel() override;

	/// <summary>
	/// Read Only: Vector of SymmetricKeySize containers, containing legal cipher input key sizes
	/// </summary>
	const std::vector<SymmetricKeySize> &LegalKeySizes() override;

	/// <summary>
	/// Read Only: The stream ciphers formal implementation name
	/// </summary>
	const std::string Name() override;

	/// <summary>
	/// Read Only: The current value of the nonce counter array.
	/// </summary>
	const std::vector<byte> Nonce() override;

	/// <summary>
	/// Read Only: Parallel block size; the byte-size of the input/output data arrays passed to a transform that trigger parallel processing.
	/// <para>This value can be changed through the ParallelProfile class, but must be a multiple of the ParallelMinimumSize().</para>
	/// </summary>
	const size_t ParallelBlockSize() override;

	/// <summary>
	/// Read/Write: Parallel and SIMD capability flags and recommended sizes.
	/// <para>The maximum number of threads allocated when using multi-threaded processing can be set with the ParallelMaxDegree() property.
	/// The ParallelBlockSize() property is auto-calculated, but can be changed; the value must be evenly divisible by ParallelMinimumSize().
	/// Changes to these values must be made before the Initialize(bool, ISymmetricKey) function is called.</para>
	/// </summary>
	ParallelOptions &ParallelProfile() override;

	/// <summary>
	/// Read Only: The current standard-vector MAC tag value
	/// </summary>
	const std::vector<byte> Tag() override;

	/// <summary>
	/// Copies the internal MAC tag to a secure-vector
	/// </summary>
	/// 
	/// <param name="Output">The secure-vector receiving the MAC code</param>
	const void Tag(SecureVector<byte> &Output) override;

	/// <summary>
	/// Read Only: The legal MAC tag length in bytes
	/// </summary>
	const size_t TagSize() override;

	//~~~Public Functions~~~//

	/// <summary>
	/// Initialize the cipher with an ISymmetricKey key container.
	/// <para>If authentication is enabled, setting the Encryption parameter to false will decrypt and authenticate a ciphertext stream.
	/// Authentication on a decrypted stream is performed automatically; failure will throw a CryptoAuthenticationFailure exception.
	/// If encryption and authentication are set to true, the MAC code is appended to the cipher-text array.</para>
	/// </summary>
	/// 
	/// <param name="Encryption">Using Encryption or Decryption mode</param>
	/// <param name="Parameters">Cipher key structure, containing cipher key, nonce, and optional info vectors</param>
	///
	/// <exception cref="CryptoSymmetricException">Thrown if a null or invalid key is used</exception>
	void Initialize(bool Encryption, ISymmetricKey &Parameters) override;

	/// <summary>
	/// Set the maximum number of threads allocated when using multi-threaded processing.
	/// <para>When set to zero, thread count is set automatically. If set to 1, sets IsParallel() to false and runs in sequential mode. 
	/// Thread count must be an even number, and not exceed the number of processor virtual cores.</para>
	/// </summary>
	///
	/// <param name="Degree">The number of threads to allocate</param>
	/// 
	/// <exception cref="CryptoCipherModeException">Thrown if the degree parameter is invalid</exception>
	void ParallelMaxDegree(size_t Degree) override;

	/// <summary>
	/// Saves the internal state of the cipher to a secure vector.
	/// <para>The Serialize function can store the internal state of the cipher at the time it is invoked.
	/// The cipher instance can be reinitialized through a constructor option, without the need to re-call the Initialize function and associated key-schedule functions.
	/// This is useful in situations where the cipher is required intermitantly, and the entire state can be stored rather than just the key and nonce.</para>
	/// </summary>
	///
	/// <returns>The serialized cipher state</returns>
	SecureVector<byte> Serialize();

	/// <summary>
	/// Add additional data to the message authentication code generator.  
	/// <para>Must be called after Initialize(bool, ISymmetricKey), and can then be called before or after a stream segment has been processed.</para>
	/// </summary>
	/// 
	/// <param name="Input">The input vector of bytes to process</param>
	/// <param name="Offset">The starting offset within the input vector</param>
	/// <param name="Length">The number of bytes to process</param>
	///
	/// <exception cref="CryptoSymmetricException">Thrown if the cipher is not initialized</exception>
	void SetAssociatedData(const std::vector<byte> &Input, size_t Offset, size_t Length) override;

	/// <summary>
	/// Encrypt/Decrypt a vector of bytes with offset and length parameters.
	/// <para>Initialize(bool, ISymmetricKey) must be called before this method can be used. 
	///	In authenticated encryption mode, the MAC code is automatically appended to the output stream at the end of the cipher-text, the output array must be long enough to accommodate this TagSize() code.
	/// In decryption mode, this code is checked before the stream is decrypted, if the authentication fails a CryptoAuthenticationFailure exception is thrown.</para>
	/// </summary>
	/// 
	/// <param name="Input">The input vector of bytes to transform</param>
	/// <param name="InOffset">The starting offset within the input vector</param>
	/// <param name="Output">The output vector of transformed bytes</param>
	/// <param name="OutOffset">The starting offset within the output vector</param>
	/// <param name="Length">Number of bytes to process</param>
	///
	/// <exception cref="CryptoAuthenticationFailure">Thrown during decryption if the the ciphertext fails authentication</exception>
	void Transform(const std::vector<byte> &Input, size_t InOffset, std::vector<byte> &Output, size_t OutOffset, size_t Length) override;

private:

	static void Finalize(std::unique_ptr<AcsState> &State, std::unique_ptr<IMac> &Authenticator);
	void Generate(std::vector<byte> &Output, size_t OutOffset, size_t Length, std::vector<byte> &Counter);
#if defined(CEX_HAS_AVX512)
	__m512i Load256To512(__m256i &A, __m256i &B);
#endif
	void Process(const std::vector<byte> &Input, size_t InOffset, std::vector<byte> &Output, size_t OutOffset, size_t Length);
	void ProcessParallel(const std::vector<byte> &Input, size_t InOffset, std::vector<byte> &Output, size_t OutOffset, size_t Length);
	void ProcessSequential(const std::vector<byte> &Input, size_t InOffset, std::vector<byte> &Output, size_t OutOffset, size_t Length);
	void Reset();
#if defined(CEX_HAS_AVX512)
	__m512i Shuffle512(const __m512i &Value, const __m512i &Mask);
#endif
#if defined(CEX_EXTENDED_AESNI)
	__m256i Shuffle256(const __m256i &Value, const __m256i &Mask);
#endif
	void Transform256(const std::vector<byte> &Input, size_t InOffset, std::vector<byte> &Output, size_t OutOffset);
	void Transform512(const std::vector<byte> &Input, size_t InOffset, std::vector<byte> &Output, size_t OutOffset);
};

NAMESPACE_STREAMEND
#endif
